src/runtime-rs/crates/resource/src/rootfs/virtual_volume.rs

```diff
@@ -0,0 +1,248 @@
+/// Otherwise, it attempts to find the image name using the appropriate Kubernetes
+/// annotation key.
+pub fn get_image_reference(spec_annotations: &HashMap<String, String>) -> Result<&str> {
+    info!(
+        sl!(),
+        "get image reference from spec annotation: {:?}", spec_annotations
+    );
+    for &key in CRI_CONTAINER_TYPE_KEY_LIST {
+        if let Some(type_value) = spec_annotations.get(key) {
+            if let Ok(container_type) = ContainerType::from_str(type_value) {
+                return match container_type {
+                    ContainerType::PodSandbox => Ok("pause"),
+                    _ => {
+                        let image_name_key = if key == annotations::crio::CONTAINER_TYPE_LABEL_KEY {
+                            KUBERNETES_CRIO_IMAGE_NAME
+                        } else {
+                            KUBERNETES_CRI_IMAGE_NAME
+                        };
+
+                        spec_annotations
+                            .get(image_name_key)
+                            .map(AsRef::as_ref)
+                            .ok_or_else(|| anyhow!("get image reference failed"))
+                    }
+                };
+            }
+        }
+    }
+
+    Err(anyhow!("no target image reference found"))
+}
+
+/// Handles storage configuration for KataVirtualVolume based on its type.
+/// Specifically processes `KATA_VIRTUAL_VOLUME_IMAGE_GUEST_PULL` type.
+fn handle_virtual_volume_storage(
+    cid: &str,
+    annotations: &HashMap<String, String>,
+    virt_volume: &KataVirtualVolume,
+) -> Result<agent::Storage> {
+    if virt_volume.volume_type.as_str() == KATA_VIRTUAL_VOLUME_IMAGE_GUEST_PULL {
+        let mut vol = handle_image_guest_pull_virtual_volume(annotations, virt_volume)
+            .context("handle image guest pull failed")?;
+        vol.fs_type = KATA_VIRTUAL_VOLUME_TYPE_OVERLAY_FS.to_string();
+        vol.mount_point = Path::new(KATA_GUEST_ROOT_SHARED_FS)
+            .join(cid)
+            .join("rootfs")
+            .display()
+            .to_string();
+
+        Ok(vol)
+    } else {
+        Err(anyhow!("Error in handle_virtual_volume_storage"))
+    }
+}
+
+/// Processes a virtual volume specifically for image guest-pull scenarios.
+/// It enriches the volume info with image reference and metadata, then serializes it into agent.Storage.
+fn handle_image_guest_pull_virtual_volume(
+    annotations: &HashMap<String, String>,
+    virtual_volume_info: &KataVirtualVolume,
+) -> Result<agent::Storage> {
+    let container_annotations = annotations.clone();
+    let image_ref = get_image_reference(annotations).context("get image reference failed.")?;
+    let mut virtual_volume_info = virtual_volume_info.clone();
+    virtual_volume_info.source = image_ref.to_owned();
+
+    // Merge metadata
+    for (k, v) in container_annotations {
+        if let Some(ref mut image_pull) = virtual_volume_info.image_pull {
+            image_pull.metadata.insert(k, v);
+        }
+    }
```

Comment:
Do I understand correctly that this means "if `image_pull` is set, copy all annotations"? In that case, I'd find it more natural to move the for loop inside the if body.